File: medium/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return sumTwoNumber(l1, l2, 0);
    }
    
    public ListNode sumTwoNumber(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null && carry == 0) {
            return null;
        }
        int total = 0;
        if (l1 != null) {
            total += l1.val;
        }
        if (l2 != null) {
            total += l2.val;
        }
        total += carry;
        ListNode val = new ListNode((total) % 10);
        val.next = sumTwoNumber(l1 != null ? l1.next : null, l2 != null ? l2.next : null, total / 10);
        return val;
        
    }
}

File: medium/add-two-numbers.js
var addTwoNumbers = function(l1, l2) {
    return recurse(l1, l2, 0);
};

const recurse = function (l1, l2, carry) {
    let total = carry;
   if (l1) {
       total += l1.val;
   }
   if (l2) {
       total += l2.val;
   }
    const newNode = new ListNode(total % 10);
    const extra =  parseInt(total / 10);
    
   if ((l1 && l1.next) || (l2 && l2.next) || extra) {
      newNode.next = recurse(l1 ? l1.next : null, l2 ? l2.next : null, extra);
    }
    return newNode;
};